graph.go

```diff
@@ -4,193 +4,166 @@
-	Neighbors(Node) []Node
-
-	// EdgeBetween returns an edge between node and neighbor such that
-	// From is one argument and To is the other. If no
-	// such edge exists, this function returns nil.
-	EdgeBetween(node, neighbor Node) Edge
-}
+	// Has returns whether the node exists within the graph.
+	Has(Node) bool
 
-// Directed graphs are characterized by having seperable Heads and Tails in their edges.
-// That is, if node1 goes to node2, that does not necessarily imply that node2 goes to node1.
-//
-// While it's possible for a directed graph to have fully reciprocal edges (i.e. the graph is
-// symmetric) -- it is not required to be. The graph is also required to implement Graph
-// because in many cases it can be useful to know all neighbors regardless of direction.
-type DirectedGraph interface {
-	Graph
-	// Successors gives the nodes connected by OUTBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Predecessors.
-	Successors(Node) []Node
-
-	// EdgeTo returns an edge between node and successor such that
-	// From returns node and To returns successor, if no
-	// such edge exists, this function returns nil.
-	EdgeTo(node, successor Node) Edge
-
-	// Predecessors gives the nodes connected by INBOUND edges.
-	// If the graph is an undirected graph, this set is equal to Successors.
-	Predecessors(Node) []Node
-}
+	// Nodes returns all the nodes in the graph.
+	Nodes() []Node
 
-// Returns all undirected edges in the graph
-type EdgeLister interface {
-	EdgeList() []Edge
-}
+	// From returns all nodes that can be reached from
+	// the given node.
+	From(Node) []Node
 
-type EdgeListGraph interface {
-	Graph
-	EdgeLister
-}
+	// HasEdge returns whether an edge exists between
+	// nodes x and y without considering direction.
+	HasEdge(x, y Node) bool
 
-// Returns all directed edges in the graph.
-type DirectedEdgeLister interface {
-	DirectedEdgeList() []Edge
+	// Edge returns the edge between nodes u and v when
+	// the nodes returned by From(u) include v.
+	Edge(u, v Node) Edge
 }
 
-type DirectedEdgeListGraph interface {
+// Undirected is an undirected graph.
+type Undirected interface {
 	Graph
-	DirectedEdgeLister
-}
 
-// A Graph that implements Coster has an actual cost between adjacent nodes, also known as a
-// weighted graph. If a graph implements coster and a function needs to read cost (e.g. A*),
-// this function will take precedence over the Uniform Cost function (all weights are 1) if "nil"
-// is passed in for the function argument.
-//
-// If the argument is nil, or the edge is invalid for some reason, this should return math.Inf(1)
-type Coster interface {
-	Cost(Edge) float64
+	// EdgeBetween returns the edge between nodes u and v.
+	EdgeBetween(u, v Node) Edge
 }
 
-type CostGraph interface {
-	Coster
+// Directed is a directed graph.
+type Directed interface {
 	Graph
+
+	// EdgeFromTo returns the edge leading from u to v.
+	EdgeFromTo(u, v Node) Edge
+
```

Comment:
I've dropped EdgeFromTo (it is redundant since Edge does exactly the same thing in a directed graph), so HasEdgeFromTo can trade its place.

It could be argued that EdgeBetween is also redundant. At the moment it is here as an interface differentiation method, but I think in the future we could make it semantically different; Edge is defined as a directed query (even in undirected graphs), while EdgeBetween is not. This is subtly different, because it is conceivable that we could return the edge ordered as queried for the Edge call, or arbitrarily ordered in the EdgeBetween case. This is not possible at the moment since the graph cannot reorder the Edge, but if we make that possible this difference could be implemented. Whether it's worth it is another matter (I can see reasons, but they are not hugely compelling).